src/v/bytes/details/io_fragment.h

```diff
@@ -94,14 +102,273 @@ class io_fragment {
+        // NOLINTNEXTLINE(hicpp-explicit-conversions)
+        operator iter<true, Forward>() const {
+            check_generation();
+            return {_container, _current};
+        }
+
+        reference operator*() const {
+            check_generation();
+            return *_current;
+        }
+        pointer operator->() const {
+            check_generation();
+            return _current;
+        }
+        iter& operator++() {
+            if constexpr (Forward) {
+                _current = _current->_next;
+            } else {
+                _current = _current->_prev;
+            }
+            return *this;
+        }
+        iter operator++(int) {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+        iter& operator--() {
+            if constexpr (Forward) {
+                if (_current == nullptr) {
+                    _current = _container->_tail;
+                } else {
+                    _current = _current->_prev;
+                }
+            } else {
+                if (_current == nullptr) {
+                    _current = _container->_head;
+                } else {
+                    _current = _current->_next;
+                }
+            }
+            return *this;
+        }
+        iter operator--(int) {
+            auto tmp = *this;
+            --*this;
+            return tmp;
+        }
+        bool operator==(const iter& o) const { return _current == o._current; }
+        bool operator!=(const iter& o) const { return _current != o._current; }
+
+    private:
+        friend class io_fragment_list;
+
+        inline void check_generation() const {
+#ifndef NDEBUG
+            vassert(
+              _container->_generation == _my_generation,
+              "Attempting to use an invalidated iterator. The corresponding "
+              "iobuf has been mutated since this iterator was constructed.");
+#endif
+        }
+
+        pointer _current = nullptr;
+#ifndef NDEBUG
+        size_t _my_generation = 0;
+#endif
+        container_type* _container = nullptr;
+    };
+
+public:
+    io_fragment_list() noexcept = default;
+    io_fragment_list(io_fragment_list&& o) noexcept
+      : _head(std::exchange(o._head, nullptr))
+      , _tail(std::exchange(o._tail, nullptr)) {}
+    io_fragment_list(const io_fragment_list&) = delete;
+    io_fragment_list& operator=(io_fragment_list&&) = delete;
+    io_fragment_list& operator=(const io_fragment_list&) = delete;
+    ~io_fragment_list() noexcept {
+        vassert(
+          _head == nullptr && _tail == nullptr,
+          "io_fragment_list cleanup failure and possible memory leak");
+    };
+
+    using iterator = iter<false, true>;
+    using reverse_iterator = iter<false, false>;
+    using const_iterator = iter<true, true>;
+
+    bool empty() const {
+        bool e = _head == nullptr;
+        assert(e == (_tail == nullptr));
+        return e;
+    }
+
+    io_fragment& front() {
```

Comment:
This override doesn't seem needed? The const-qualified version below is identical.

Reply:
`front` is only used in non-const contexts so I've just removed the const override.